File: Geek Assistant (Android AIO)/Modules/Android/Function/madb_Functions.vb
Imports Managed

Module madb_Functions

#Region "madb assited"

    ''' <summary>
    ''' Start adb server at $"{GA_tools}\adb.exe" or skip if already running
    ''' </summary>
    Public Sub madbStart()
        'Dim madb = 
        Adb.AndroidDebugBridge.CreateBridge($"{GA_tools}\adb.exe", False)
        'madb.Start()
    End Sub

    ''' <summary>
    ''' Terminate adb server 
    ''' </summary>
    Public Sub madbStop()
        Adb.AndroidDebugBridge.Terminate()
    End Sub

    ''' <summary>
    ''' Count how many devices are currently connected
    ''' </summary>
    ''' <returns>Get the count of madb_DeviceList() As Integer ( /List(Of Adb.Device)/ .Count)</returns>
    Public Function madb_GetDeviceCount() As Integer
        madbStart() 'Failsafe
        Return madb_GetListOfDevice.Count
    End Function

    ''' <summary>
    ''' Get a list of the devices connected as List(Of Adb.Device)
    ''' </summary>
    ''' <returns>Get a list of the devices connected as List(Of Adb.Device)</returns>
    Public Function madb_GetListOfDevice() As List(Of Adb.Device)
        madbStart() 'Failsafe
        Dim DeviceList As List(Of Adb.Device) = Adb.AdbHelper.Instance.GetDevices(Adb.AndroidDebugBridge.SocketAddress)
        Return DeviceList
    End Function

    ''' <summary>
    ''' Check if any connected device serial# matches the input serial# and return as Adb.Device
    ''' </summary>
    ''' <param name="serial">Serial number to check against the connected devices</param>
    ''' <returns>Get the Adb.Device that matches the serial number</returns>
    Public Function madb_GetDeviceFromSerial(serial As String) As Adb.Device
        madbStart() 'Failsafe
        Dim devices As List(Of Adb.Device) = Adb.AdbHelper.Instance.GetDevices(Adb.AndroidDebugBridge.SocketAddress)
        For Each dev As Adb.Device In devices
            If dev.SerialNumber = serial Then Return dev
        Next
        Return Nothing
    End Function
    ''' <summary>
    ''' Get the state of the first device connected
    ''' </summary>
    ''' <returns></returns>
    Public Function madb_GetDeviceState() As Integer
        madbStart() 'Failsafe
        Return madb_GetListOfDevice(0).State
    End Function

#End Region

End Module

File: Geek Assistant (Android AIO)/Modules/Android/Function/madb_Functions.cs
using System.Collections.Generic;
using Managed;

internal static partial class madb {

    /// <summary>
    /// Get adb server at $"{GA_tools}\adb.exe"
    /// </summary>
    /// <param name="forceNewBridge">Kill previous server if true</param>
    /// <returns>Managed.Adb.AndroidDebugBridge</returns>
    public static Managed.Adb.AndroidDebugBridge madbBridge(bool forceNewBridge = false) {
        return Managed.Adb.AndroidDebugBridge.CreateBridge($@"{c.GA_tools}\adb.exe", forceNewBridge); // .Start()
    }
    /// <summary>
    /// Start adb server at $"{GA_tools}\adb.exe"
    /// </summary>
    /// <param name="forceNewBridge">Kill previous server if true</param>
    public static void madbBridgeStart(bool forceNewBridge = false) {
        madbBridge(forceNewBridge).Start(); // .Start()
    }
    /// <summary>
    /// Terminate madb
    /// </summary>
    public static void madbStop() {
        Managed.Adb.AndroidDebugBridge.Terminate();
    }

    /// <summary>
    /// Count how many devices are currently connected
    /// </summary>
    /// <returns>Get the count of madb_DeviceList() </returns>
    public static int GetDeviceCount() {
        madbBridge(); // Failsafe 
        return GetListOfDevice().Count;
    }

    /// <returns>Get a list of the devices connected as List(Of Managed.Adb.Device)</returns>
    public static List<Managed.Adb.Device> GetListOfDevice() {
        madbBridge(); // Failsafe  
        return Managed.Adb.AdbHelper.Instance.GetDevices(Managed.Adb.AndroidDebugBridge.SocketAddress);
    }

    /// <summary>
    /// Check if any connected device serial# matches the input serial# and return as Managed.Adb.Device
    /// </summary>
    /// <param name="serial">Serial number to check against the connected devices</param>
    /// <returns>Get the Managed.Adb.Device that matches the serial number</returns>
    public static Managed.Adb.Device GetDeviceFromSerial(string serial) {
        madbBridge(); // Failsafe
        List<Managed.Adb.Device> devices = Managed.Adb.AdbHelper.Instance.GetDevices(Managed.Adb.AndroidDebugBridge.SocketAddress);
        foreach (Managed.Adb.Device dev in devices) {
            if (dev.SerialNumber == serial)
                return dev;
        }

        return default;
    }

    /// <summary>
    /// Get the state of the first device connected
    /// </summary>
    /// <returns>(Integer): 0 recovery | 1 bootloader | 2 offline | 3 online | 4 download | 5 unknown</returns>
    public static int GetDeviceState() {
        madbBridge(); // Failsafe
        return (int)GetListOfDevice()[0].State;
    }
    /// <summary>
    /// Converts madb_GetDeviceState() to the corresponding string
    /// </summary>
    /// <returns>(String): 0 recovery | 1 bootloader | 2 offline | 3 online | 4 download | 5 unknown</returns>
    public static string Convert_DeviceState_IntToString() {
        string result = "";
        switch (GetDeviceState()) {

        case 0: {
            result = "recovery";
            break;
        }

        case 1: {
            result = "bootloader";
            break;
        }

        case 2: {
            result = "offline";
            break;
        }

        case 3: {
            result = "online";
            break;
        }

        case 4: {
            result = "download";
            break;
        }

        case 5: {
            result = "unknown";
            break;
        }
        }

        return result;
    }

    /// <param name="ErrorCode_init">Error Beginning "(XX)-xx"</param>
    /// <returns>True if Device is can SU</returns>
    public static bool madb_IsRooted(string ErrorCode_init) {
        var dev = GetListOfDevice()[0];
        if (!dev.CanSU()) {
            c.ErrorInfo = ($"{ErrorCode_init}-Xsu", 1, $"Your device is not rooted.\n > Process aborted."); // X su (Device cannot run su)
            return false;
        }

        return true;
    }
}